### PR TITLE
Add readonly to structs

### PR DESCRIPTION
**Pull request by teinarss**

### CODE REVIEW

File: OpenRA.Game/WPos.cs [hunk 1]
```diff
@@ -29,8 +29,8 @@ public struct WPos : IScriptBindable, ILuaAdditionBinding, ILuaSubtractionBindin
 
 		public static explicit operator WVec(WPos a) { return new WVec(a.X, a.Y, a.Z); }
 
-		public static WPos operator +(WPos a, WVec b) { return new WPos(a.X + b.X, a.Y + b.Y, a.Z + b.Z); }
-		public static WPos operator -(WPos a, WVec b) { return new WPos(a.X - b.X, a.Y - b.Y, a.Z - b.Z); }
+		public static WPos operator +(WPos a, in WVec b) { return new WPos(a.X + b.X, a.Y + b.Y, a.Z + b.Z); }
```
Comment (anvilvapre) [inline OpenRA.Game/WPos.cs]: I suspect a very large performance improvement. Unless optimization in the past already took care of this.

Comment (anvilvapre) [reply OpenRA.Game/WPos.cs]: Both can't be passed by reference?
Wouldn't it then be better to create a non static `operator +` also (if possible in c#).

File: OpenRA.Game/Graphics/TargetLineRenderable.cs [hunk 2]
```diff
@@ -35,7 +35,13 @@ public TargetLineRenderable(IEnumerable<WPos> waypoints, Color color, int width
 		public bool IsDecoration => true;
 
 		public IRenderable WithZOffset(int newOffset) { return new TargetLineRenderable(waypoints, color); }
-		public IRenderable OffsetBy(WVec vec) { return new TargetLineRenderable(waypoints.Select(w => w + vec), color); }
+
+		public IRenderable OffsetBy(in WVec vec)
+		{
+			var vec1 = vec;
```
Comment (anvilvapre) [inline OpenRA.Game/Graphics/TargetLineRenderable.cs]: Doesn't this defy the pass by reference? But assume needed for the LINQ statement.

**Comment by anvilvapre** [comment]
`CellLayer Index[CPos]` could benefit from 'in'. Used in many critical parts.

**Comment by teinarss** [comment]
> 
> 
> `CellLayer Index[CPos]` could benefit from 'in'. Used in many critical parts.

@anvilvapre It is only beneficial to add the in modifier to  structs that are larger than `intptr`

File: OpenRA.Game/Graphics/TargetLineRenderable.cs (hunk 2 above)
Comment (teinarss) [reply OpenRA.Game/Graphics/TargetLineRenderable.cs]: Correct, you cant use the `in` modifier in lambdas.

Comment (pchote) [inline OpenRA.Game/Graphics/TargetLineRenderable.cs]: Can we please add the comment and naming we used in #18544?

```csharp
// Lambdas can't use 'in' variables, so capture a copy for later
var offset = vec;
```

File: OpenRA.Mods.Common/Graphics/ContrailRenderable.cs [hunk 3]
```diff
@@ -50,7 +50,13 @@ public ContrailRenderable(World world, Color color, WDist width, int length, int
 		public bool IsDecoration => true;
 
 		public IRenderable WithZOffset(int newOffset) { return new ContrailRenderable(world, (WPos[])trail.Clone(), width, next, length, skip, color, newOffset); }
-		public IRenderable OffsetBy(WVec vec) { return new ContrailRenderable(world, trail.Select(pos => pos + vec).ToArray(), width, next, length, skip, color, zOffset); }
+
+		public IRenderable OffsetBy(in WVec vec)
+		{
+			var vec1 = vec;
```
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/ContrailRenderable.cs]: and here

File: OpenRA.Mods.Common/Graphics/PolygonAnnotationRenderable.cs [hunk 4]
```diff
@@ -35,7 +35,13 @@ public PolygonAnnotationRenderable(WPos[] vertices, WPos effectivePos, int width
 		public bool IsDecoration => true;
 
 		public IRenderable WithZOffset(int newOffset) { return new PolygonAnnotationRenderable(vertices, effectivePos, width, color); }
-		public IRenderable OffsetBy(WVec vec) { return new PolygonAnnotationRenderable(vertices.Select(v => v + vec).ToArray(), effectivePos + vec, width, color); }
+
+		public IRenderable OffsetBy(in WVec vec)
+		{
+			var vec1 = vec;
```
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/PolygonAnnotationRenderable.cs]: 👆

**Comment by teinarss** [comment]
Updated.

**Review by pchote — APPROVED** [review]

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2259475ff1e3ce0f818821bb82eb208b0fea8960...d6d0a8229bc4432de3418a1a5dd28eeffea41a8a)

**Pull request merged**